// @hsui/plugin-micro-app/lib/app.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;
var _fsExtra = require("fs-extra");
var _path = _interopRequireDefault(require("path"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const _require = require('../package.json'),
  version = _require.version;
function _default(api, options = {}) {
  // @tmp/micro-app
  (0, _fsExtra.ensureDirSync)('./src/.hui/micro-app');

  // meta
  api.writeTmpFile('micro-app/meta.js', `export const version = "${version}"`);
  api.addEntryFileAhead('app', '@tmp/micro-app/setPublicPath.js');
  (0, _fsExtra.copyFileSync)(_path.default.resolve(__dirname, '../template/setPublicPath.js'), './src/.hui/micro-app/setPublicPath.js');
  api.addEntryImports(() => [{
    source: '@hsui/core',
    specifiers: ['hCoreReference', 'singleSpaVue']
  }]);
  api.addHuiExports([{
    specifiers: '*',
    source: '@tmp/micro-app/frameEmitter.js'
  }, {
    specifiers: 'singleSpaVue',
    source: '@tmp/micro-app/singleSpaVue.js'
  }]);
  (0, _fsExtra.copyFileSync)(_path.default.resolve(__dirname, '../template/frameEmitter.js'), './src/.hui/micro-app/frameEmitter.js');
  (0, _fsExtra.copyFileSync)(_path.default.resolve(__dirname, '../template/singleSpaVue.js'), './src/.hui/micro-app/singleSpaVue.js');
  if (options.vuex) {
    api.addEntryImports(() => [{
      source: '@hsui/core',
      specifiers: ['frameEmitter', 'store']
    }]);
    api.addEntryCodeAhead(() => `
if (frameEmitter) {
  frameEmitter.on("register-root-module", ({ root }) => {
    if (!store.hasModule("root")) {
      store.registerModule("root", { ...root, namespaced: true });
    }
  });

  frameEmitter.on("update-root-module", ({ mutation, root }) => {
    if (!store.hasModule("root")) {
      store.registerModule("root", { ...root, namespaced: true });
    }
    const { type, payload } = mutation;
    store.commit(type, payload);
  });

  frameEmitter.trigger("app-store-ready");
}`);
  }

  // just a compatibly handle
  if (options.hasOwnProperty('destoryRootInstanceAtUnmount')) {
    options.destroyRootInstanceAtUnmount = options.destoryRootInstanceAtUnmount;
  }
  api.addEntryCode(() => `
const { Model, options: appOptions } = hCoreReference;
export const { bootstrap, mount, unmount } = singleSpaVue({
  Vue: Model,
  appOptions,
  destroyRootInstanceAtUnmount: ${options.hasOwnProperty('destroyRootInstanceAtUnmount') ? options.destroyRootInstanceAtUnmount : true},
});
`);
  api.chainWebpack(webpackConfig => {
    webpackConfig.output.library(`HUI_MICRO_APP_${options.name}`).libraryTarget('global').jsonpFunction(`webpackJsonp_${options.name}`);
    const isProd = process.env.HUI_CLI_ENV === 'production';
    const inlineLimit = typeof options.inlineLimit === 'undefined' ? 4096 : options.inlineLimit;
    const genUrlLoaderOptions = dir => {
      const publicPath = options.publicPath || options.base || '/';
      return {
        limit: inlineLimit,
        esModule: false,
        // use explicit fallback to avoid regression in url-loader>=1.1.0
        fallback: {
          loader: 'file-loader',
          options: {
            name: `${dir}/[name]${options.filenameHashing ? '.[hash:8]' : ''}.[ext]`,
            esModule: false,
            publicPath: isProd ? publicPath : `http://${process.env.HUI_CLI_DEV_SERVER_IP_ADDRESS || 'localhost'}:${process.env.HUI_CLI_DEV_SERVER_PORT}${publicPath}` // https://qiankun.umijs.org/zh/faq#%E4%B8%BA%E4%BB%80%E4%B9%88%E5%BE%AE%E5%BA%94%E7%94%A8%E5%8A%A0%E8%BD%BD%E7%9A%84%E8%B5%84%E6%BA%90%E4%BC%9A-404%EF%BC%9F
          }
        }
      };
    };

    // static assets -----------------------------------------------------------

    webpackConfig.module.rule('images').use('url-loader').loader('url-loader').options(genUrlLoaderOptions('img'));
    webpackConfig.module.rule('media').use('url-loader').loader('url-loader').options(genUrlLoaderOptions('media'));
    webpackConfig.module.rule('fonts').use('url-loader').loader('url-loader').options(genUrlLoaderOptions('fonts'));
    const webpack = api.resolveBundlerDep('webpack');
    if (!isProd) {
      webpackConfig.devtool(false);
      webpackConfig.plugin('SourceMapDevToolPlugin').use(webpack.SourceMapDevToolPlugin, [{
        namespace: options.name
      }]);
    }
  });
}
