// h_ui/dist/style/theme-chalk/gulpfile.js
const { series, src, dest } = require('gulp');
const sass = require('gulp-sass');
const autoprefixer = require('gulp-autoprefixer');
const cssmin = require('gulp-cssmin');
const rename = require('gulp-rename');

const targetPath = '../../../dist/lib';

function compile() {
  return src([
    // '../components/button.scss',
    // '../components/select.scss',
    // '../components/dropdown.scss',
    // '../components/icon.scss',
    // '../components/editgird.scss',
    '../components/singleSelect.scss',
    '../components/upload.scss',
    '../components/badge.scss',
    '../components/carousel.scss',
    '../components/collapse.scss',
    '../components/tag.scss',
    '../components/timeline.scss',
    '../components/tooltip.scss'
    '../components/simpleTable.scss'
  ])
    .pipe(sass.sync())
    .pipe(
      autoprefixer({
        browsers: ['ie > 9', 'last 2 versions'],
        cascade: false
      })
    )
    .pipe(cssmin())
    .pipe(rename(function(path) {
      // Updates the object in-place
      path.basename = path.basename.replace(/-/g,'').toLowerCase();
    }))
    .pipe(dest(targetPath + '/theme-chalk'));
}

function common() {
  return src('../common/index.scss')
    .pipe(sass.sync())
    .pipe(
      autoprefixer({
        browsers: ['ie > 9', 'last 2 versions'],
        cascade: false
      })
    )
    .pipe(cssmin())
    .pipe(dest(targetPath + '/theme-chalk/common'));
}

function copyfont() {
  return src('../icon/*.{eot,ttf,woff,svg}').pipe(dest(targetPath + '/theme-chalk/common/icon'));
}

exports.build = series(common, compile, copyfont);
